lib/puppet/util/windows/service.rb

```diff
@@ -0,0 +1,726 @@
+    # @return [QUERY_SERVICE_CONFIGW.struct] the configuration of the service
+    def service_start_type(service_name)
+      config = nil
+      open_service(service_name, SC_MANAGER_CONNECT, SERVICE_QUERY_CONFIG) do |service|
+        config = query_config(service)
+      end
+      case config[:dwStartType]
+      when SERVICE_AUTO_START
+        'SERVICE_AUTO_START'
+      when SERVICE_BOOT_START
+        'SERVICE_BOOT_START'
+      when SERVICE_SYSTEM_START
+        'SERVICE_SYSTEM_START'
+      when SERVICE_DEMAND_START
+        'SERVICE_DEMAND_START'
+      when SERVICE_DISABLED
+        'SERVICE_DISABLED'
+      else
+        raise Puppet::Error.new(_("Unknown start type '%{start_type}' for '%{service_name}'") % { start_type: status[:dwStartType].to_s, service_name: service_name})
+      end
+    end
+    module_function :service_start_type
+
+    # Change the startup mode of a windows service
+    #
+    # @param [string] service_name the name of the service to modify
+    # @param [Int] startup_type a code corresponding to a start type for
+    #  windows service, see the "Service start type codes" section in the
+    #  Puppet::Util::Windows::Service file for the list of available codes
+    def set_startup_mode(service_name, startup_type)
+      case startup_type
+      when 'SERVICE_AUTO_START'
+        startup_code = SERVICE_AUTO_START
+      when 'SERVICE_BOOT_START'
+        startup_code = SERVICE_BOOT_START
+      when 'SERVICE_SYSTEM_START'
+        startup_code = SERVICE_SYSTEM_START
+      when 'SERVICE_DEMAND_START'
+        startup_code = SERVICE_DEMAND_START
+      when 'SERVICE_DISABLED'
+        startup_code = SERVICE_DISABLED
+      else
+        raise Puppet::Error.new(_("Unknown start type %{start_type}") % {startup_type: startup_type.to_s})
+      end
+      open_service(service_name, SC_MANAGER_CONNECT, SERVICE_CHANGE_CONFIG) do |service|
+        # Currently the only thing puppet's API can really manage
+        # in this list is dwStartType (the third param). Thus no
+        # generic function was written to make use of all the params
+        # since the API as-is couldn't use them anyway
+        if ChangeServiceConfigW(
+          service,
+          SERVICE_NO_CHANGE,
+          startup_code,
+          SERVICE_NO_CHANGE,
+          nil,
+          nil,
+          nil,
+          nil,
+          nil,
+          nil,
+          nil
+        ) == FFI::WIN32_FALSE
+          raise Puppet::Util::Windows::Error.new(_("Failed to update service configuration"))
+        end
+      end
+    end
+    module_function :set_startup_mode
+
+    # enumerate over all services in all states and return them as a hash
+    #
+    # @return [Hash] a hash containing services:
+    #   { 'service name' => {
+    #                         'display_name' => 'display name',
+    #                         'service_status_process' => SERVICE_STATUS_PROCESS struct
+    #                       }
+    #   }
+    def enum_services
```

Comment:
Could this be renamed to `services`? Retrieving the services is what it's doing.